File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}
=begin
last_seen = {
    p => 0,
    w => 2,
    k => 3,
    e => 4,
    
}
max_length = 3
l = 2
r = 4
s = "pwwkew"
          ^
=end
def length_of_longest_substring(s)
    
    last_seen = {}
    max_length = l = 0
    
    s.chars.each_with_index do |c, r|
        if last_seen.key?(c)
            l = [l, last_seen[c] + 1].max
        end
        max_length = [max_length, r - l + 1].max
        last_seen[c] = r
    end
    max_length
end

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
"""
max_length = 3
window = {a = 3, 
          b = 2, 
          c = 5}
move_l = max(l, window[s[l]] + 1)

    0 1 2 3 4 5 6 7
s = a b c a b c b b
              l
                 r

"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = l = 0
        r = len(s) - 1
        window = {}
        
        for r, c in enumerate(s):
            if c in window:
                l = max(l, window[c] + 1)
            window[c] = r
            res = max(res, r - l + 1)
        
        return res
        